{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of Contents\n",
    "- <a href='#1.0'>Section 1 - Introduction</a>\n",
    "    - <a href='#1.1'>Section 1.1 - Downloading our Data</a>\n",
    "- <a href='#2.0'>Section 2 - Running BLAST</a>\n",
    "    - <a href='#2.1'>Section 2.1 - Setting up the Database</a>\n",
    "    - <a href='#2.2'>Section 2.2 - A Brief Look Into the BLAST Algorithm</a>\n",
    "    - <a href='#2.3'>Section 2.3 - Running the BLAST Search</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial 1 - Command Line BLAST\n",
    "<img src='img/tut01/ncbi.png'></img>\n",
    "## <a id='1.0'>Section 1 - Introduction</a>\n",
    "\n",
    "If you were ask any bioinformatician about what they think the most important tool they use is, they would all likely say BLAST.\n",
    "\n",
    "But what does BLAST stand for?\n",
    "\n",
    "$B$asic\n",
    "<br>\n",
    "$L$ocal\n",
    "<br>\n",
    "$A$lignment\n",
    "<br>\n",
    "$S$earch\n",
    "<br>\n",
    "$T$ool\n",
    "\n",
    "It is used to search for either Nucleotide (DNA/RNA) or Amino Acid sequences against the entire NCBI database. This is no trivial task since there are billions of sequences currently uploaded to the database. In this tutorial, we will go over the basic architecture behind the NCBI database while going over how the BLAST algorithm works."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='1.1'>Section 1.1 - Downloading our Data</a>\n",
    "\n",
    "To save a bit of time, we're not going to run BLAST searches against the entire NCBI database. We're going to download both the genome and proteome for a strain of E. coli called O157H7. The links can be found here....\n",
    "\n",
    "Genome:\n",
    "ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Bacteria/Escherichia_coli_O157_H7_uid57781/NC_002127.fna\n",
    "\n",
    "Proteome:\n",
    "ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Bacteria/Escherichia_coli_O157_H7_uid57781/NC_002127.faa\n",
    "\n",
    "When working on a project of your own, it's often faster to download a few of the important sequences you'll be using so that you can use the Commmand Line to run the BLAST algorithm. To do that, we will be using the Command Line function `wget` to download our data. To see all of the options we could use for wget, run the cell below..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GNU Wget 1.19.4, a non-interactive network retriever.\n",
      "Usage: wget [OPTION]... [URL]...\n",
      "\n",
      "Mandatory arguments to long options are mandatory for short options too.\n",
      "\n",
      "Startup:\n",
      "  -V,  --version                   display the version of Wget and exit\n",
      "  -h,  --help                      print this help\n",
      "  -b,  --background                go to background after startup\n",
      "  -e,  --execute=COMMAND           execute a `.wgetrc'-style command\n",
      "\n",
      "Logging and input file:\n",
      "  -o,  --output-file=FILE          log messages to FILE\n",
      "  -a,  --append-output=FILE        append messages to FILE\n",
      "  -d,  --debug                     print lots of debugging information\n",
      "  -q,  --quiet                     quiet (no output)\n",
      "  -v,  --verbose                   be verbose (this is the default)\n",
      "  -nv, --no-verbose                turn off verboseness, without being quiet\n",
      "       --report-speed=TYPE         output bandwidth as TYPE.  TYPE can be bits\n",
      "  -i,  --input-file=FILE           download URLs found in local or external FILE\n",
      "  -F,  --force-html                treat input file as HTML\n",
      "  -B,  --base=URL                  resolves HTML input-file links (-i -F)\n",
      "                                     relative to URL\n",
      "       --config=FILE               specify config file to use\n",
      "       --no-config                 do not read any config file\n",
      "       --rejected-log=FILE         log reasons for URL rejection to FILE\n",
      "\n",
      "Download:\n",
      "  -t,  --tries=NUMBER              set number of retries to NUMBER (0 unlimits)\n",
      "       --retry-connrefused         retry even if connection is refused\n",
      "  -O,  --output-document=FILE      write documents to FILE\n",
      "  -nc, --no-clobber                skip downloads that would download to\n",
      "                                     existing files (overwriting them)\n",
      "       --no-netrc                  don't try to obtain credentials from .netrc\n",
      "  -c,  --continue                  resume getting a partially-downloaded file\n",
      "       --start-pos=OFFSET          start downloading from zero-based position OFFSET\n",
      "       --progress=TYPE             select progress gauge type\n",
      "       --show-progress             display the progress bar in any verbosity mode\n",
      "  -N,  --timestamping              don't re-retrieve files unless newer than\n",
      "                                     local\n",
      "       --no-if-modified-since      don't use conditional if-modified-since get\n",
      "                                     requests in timestamping mode\n",
      "       --no-use-server-timestamps  don't set the local file's timestamp by\n",
      "                                     the one on the server\n",
      "  -S,  --server-response           print server response\n",
      "       --spider                    don't download anything\n",
      "  -T,  --timeout=SECONDS           set all timeout values to SECONDS\n",
      "       --dns-timeout=SECS          set the DNS lookup timeout to SECS\n",
      "       --connect-timeout=SECS      set the connect timeout to SECS\n",
      "       --read-timeout=SECS         set the read timeout to SECS\n",
      "  -w,  --wait=SECONDS              wait SECONDS between retrievals\n",
      "       --waitretry=SECONDS         wait 1..SECONDS between retries of a retrieval\n",
      "       --random-wait               wait from 0.5*WAIT...1.5*WAIT secs between retrievals\n",
      "       --no-proxy                  explicitly turn off proxy\n",
      "  -Q,  --quota=NUMBER              set retrieval quota to NUMBER\n",
      "       --bind-address=ADDRESS      bind to ADDRESS (hostname or IP) on local host\n",
      "       --limit-rate=RATE           limit download rate to RATE\n",
      "       --no-dns-cache              disable caching DNS lookups\n",
      "       --restrict-file-names=OS    restrict chars in file names to ones OS allows\n",
      "       --ignore-case               ignore case when matching files/directories\n",
      "  -4,  --inet4-only                connect only to IPv4 addresses\n",
      "  -6,  --inet6-only                connect only to IPv6 addresses\n",
      "       --prefer-family=FAMILY      connect first to addresses of specified family,\n",
      "                                     one of IPv6, IPv4, or none\n",
      "       --user=USER                 set both ftp and http user to USER\n",
      "       --password=PASS             set both ftp and http password to PASS\n",
      "       --ask-password              prompt for passwords\n",
      "       --use-askpass=COMMAND       specify credential handler for requesting \n",
      "                                     username and password.  If no COMMAND is \n",
      "                                     specified the WGET_ASKPASS or the SSH_ASKPASS \n",
      "                                     environment variable is used.\n",
      "       --no-iri                    turn off IRI support\n",
      "       --local-encoding=ENC        use ENC as the local encoding for IRIs\n",
      "       --remote-encoding=ENC       use ENC as the default remote encoding\n",
      "       --unlink                    remove file before clobber\n",
      "       --no-xattr                  turn off storage of metadata in extended file attributes\n",
      "\n",
      "Directories:\n",
      "  -nd, --no-directories            don't create directories\n",
      "  -x,  --force-directories         force creation of directories\n",
      "  -nH, --no-host-directories       don't create host directories\n",
      "       --protocol-directories      use protocol name in directories\n",
      "  -P,  --directory-prefix=PREFIX   save files to PREFIX/..\n",
      "       --cut-dirs=NUMBER           ignore NUMBER remote directory components\n",
      "\n",
      "HTTP options:\n",
      "       --http-user=USER            set http user to USER\n",
      "       --http-password=PASS        set http password to PASS\n",
      "       --no-cache                  disallow server-cached data\n",
      "       --default-page=NAME         change the default page name (normally\n",
      "                                     this is 'index.html'.)\n",
      "  -E,  --adjust-extension          save HTML/CSS documents with proper extensions\n",
      "       --ignore-length             ignore 'Content-Length' header field\n",
      "       --header=STRING             insert STRING among the headers\n",
      "       --max-redirect              maximum redirections allowed per page\n",
      "       --proxy-user=USER           set USER as proxy username\n",
      "       --proxy-password=PASS       set PASS as proxy password\n",
      "       --referer=URL               include 'Referer: URL' header in HTTP request\n",
      "       --save-headers              save the HTTP headers to file\n",
      "  -U,  --user-agent=AGENT          identify as AGENT instead of Wget/VERSION\n",
      "       --no-http-keep-alive        disable HTTP keep-alive (persistent connections)\n",
      "       --no-cookies                don't use cookies\n",
      "       --load-cookies=FILE         load cookies from FILE before session\n",
      "       --save-cookies=FILE         save cookies to FILE after session\n",
      "       --keep-session-cookies      load and save session (non-permanent) cookies\n",
      "       --post-data=STRING          use the POST method; send STRING as the data\n",
      "       --post-file=FILE            use the POST method; send contents of FILE\n",
      "       --method=HTTPMethod         use method \"HTTPMethod\" in the request\n",
      "       --body-data=STRING          send STRING as data. --method MUST be set\n",
      "       --body-file=FILE            send contents of FILE. --method MUST be set\n",
      "       --content-disposition       honor the Content-Disposition header when\n",
      "                                     choosing local file names (EXPERIMENTAL)\n",
      "       --content-on-error          output the received content on server errors\n",
      "       --auth-no-challenge         send Basic HTTP authentication information\n",
      "                                     without first waiting for the server's\n",
      "                                     challenge\n",
      "\n",
      "HTTPS (SSL/TLS) options:\n",
      "       --secure-protocol=PR        choose secure protocol, one of auto, SSLv2,\n",
      "                                     SSLv3, TLSv1, TLSv1_1, TLSv1_2 and PFS\n",
      "       --https-only                only follow secure HTTPS links\n",
      "       --no-check-certificate      don't validate the server's certificate\n",
      "       --certificate=FILE          client certificate file\n",
      "       --certificate-type=TYPE     client certificate type, PEM or DER\n",
      "       --private-key=FILE          private key file\n",
      "       --private-key-type=TYPE     private key type, PEM or DER\n",
      "       --ca-certificate=FILE       file with the bundle of CAs\n",
      "       --ca-directory=DIR          directory where hash list of CAs is stored\n",
      "       --crl-file=FILE             file with bundle of CRLs\n",
      "       --pinnedpubkey=FILE/HASHES  Public key (PEM/DER) file, or any number\n",
      "                                   of base64 encoded sha256 hashes preceded by\n",
      "                                   'sha256//' and separated by ';', to verify\n",
      "                                   peer against\n",
      "       --random-file=FILE          file with random data for seeding the SSL PRNG\n",
      "\n",
      "HSTS options:\n",
      "       --no-hsts                   disable HSTS\n",
      "       --hsts-file                 path of HSTS database (will override default)\n",
      "\n",
      "FTP options:\n",
      "       --ftp-user=USER             set ftp user to USER\n",
      "       --ftp-password=PASS         set ftp password to PASS\n",
      "       --no-remove-listing         don't remove '.listing' files\n",
      "       --no-glob                   turn off FTP file name globbing\n",
      "       --no-passive-ftp            disable the \"passive\" transfer mode\n",
      "       --preserve-permissions      preserve remote file permissions\n",
      "       --retr-symlinks             when recursing, get linked-to files (not dir)\n",
      "\n",
      "FTPS options:\n",
      "       --ftps-implicit                 use implicit FTPS (default port is 990)\n",
      "       --ftps-resume-ssl               resume the SSL/TLS session started in the control connection when\n",
      "                                         opening a data connection\n",
      "       --ftps-clear-data-connection    cipher the control channel only; all the data will be in plaintext\n",
      "       --ftps-fallback-to-ftp          fall back to FTP if FTPS is not supported in the target server\n",
      "WARC options:\n",
      "       --warc-file=FILENAME        save request/response data to a .warc.gz file\n",
      "       --warc-header=STRING        insert STRING into the warcinfo record\n",
      "       --warc-max-size=NUMBER      set maximum size of WARC files to NUMBER\n",
      "       --warc-cdx                  write CDX index files\n",
      "       --warc-dedup=FILENAME       do not store records listed in this CDX file\n",
      "       --no-warc-digests           do not calculate SHA1 digests\n",
      "       --no-warc-keep-log          do not store the log file in a WARC record\n",
      "       --warc-tempdir=DIRECTORY    location for temporary files created by the\n",
      "                                     WARC writer\n",
      "\n",
      "Recursive download:\n",
      "  -r,  --recursive                 specify recursive download\n",
      "  -l,  --level=NUMBER              maximum recursion depth (inf or 0 for infinite)\n",
      "       --delete-after              delete files locally after downloading them\n",
      "  -k,  --convert-links             make links in downloaded HTML or CSS point to\n",
      "                                     local files\n",
      "       --convert-file-only         convert the file part of the URLs only (usually known as the basename)\n",
      "       --backups=N                 before writing file X, rotate up to N backup files\n",
      "  -K,  --backup-converted          before converting file X, back up as X.orig\n",
      "  -m,  --mirror                    shortcut for -N -r -l inf --no-remove-listing\n",
      "  -p,  --page-requisites           get all images, etc. needed to display HTML page\n",
      "       --strict-comments           turn on strict (SGML) handling of HTML comments\n",
      "\n",
      "Recursive accept/reject:\n",
      "  -A,  --accept=LIST               comma-separated list of accepted extensions\n",
      "  -R,  --reject=LIST               comma-separated list of rejected extensions\n",
      "       --accept-regex=REGEX        regex matching accepted URLs\n",
      "       --reject-regex=REGEX        regex matching rejected URLs\n",
      "       --regex-type=TYPE           regex type (posix|pcre)\n",
      "  -D,  --domains=LIST              comma-separated list of accepted domains\n",
      "       --exclude-domains=LIST      comma-separated list of rejected domains\n",
      "       --follow-ftp                follow FTP links from HTML documents\n",
      "       --follow-tags=LIST          comma-separated list of followed HTML tags\n",
      "       --ignore-tags=LIST          comma-separated list of ignored HTML tags\n",
      "  -H,  --span-hosts                go to foreign hosts when recursive\n",
      "  -L,  --relative                  follow relative links only\n",
      "  -I,  --include-directories=LIST  list of allowed directories\n",
      "       --trust-server-names        use the name specified by the redirection\n",
      "                                     URL's last component\n",
      "  -X,  --exclude-directories=LIST  list of excluded directories\n",
      "  -np, --no-parent                 don't ascend to the parent directory\n",
      "\n",
      "Mail bug reports and suggestions to <bug-wget@gnu.org>\n"
     ]
    }
   ],
   "source": [
    "!wget -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now in the cell below, use one of the `wget` options to download the E. coli genome and proteome into our `data` directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2019-01-27 00:14:56--  ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Bacteria/Escherichia_coli_O157_H7_uid57781/NC_002127.fna\n",
      "           => ‘data/NC_002127.fna’\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.7, 2607:f220:41e:250::7\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.7|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/archive/old_refseq/Bacteria/Escherichia_coli_O157_H7_uid57781 ... done.\n",
      "==> SIZE NC_002127.fna ... 3454\n",
      "==> PASV ... done.    ==> RETR NC_002127.fna ... done.\n",
      "Length: 3454 (3.4K) (unauthoritative)\n",
      "\n",
      "NC_002127.fna       100%[===================>]   3.37K  --.-KB/s    in 0.02s   \n",
      "\n",
      "2019-01-27 00:14:56 (166 KB/s) - ‘data/NC_002127.fna’ saved [3454]\n",
      "\n",
      "--2019-01-27 00:14:56--  ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Bacteria/Escherichia_coli_O157_H7_uid57781/NC_002127.faa\n",
      "           => ‘data/NC_002127.faa’\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.7, 2607:f220:41e:250::7\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.7|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/archive/old_refseq/Bacteria/Escherichia_coli_O157_H7_uid57781 ... done.\n",
      "==> SIZE NC_002127.faa ... 875\n",
      "==> PASV ... done.    ==> RETR NC_002127.faa ... done.\n",
      "Length: 875 (unauthoritative)\n",
      "\n",
      "NC_002127.faa       100%[===================>]     875  --.-KB/s    in 0.02s   \n",
      "\n",
      "2019-01-27 00:14:57 (47.3 KB/s) - ‘data/NC_002127.faa’ saved [875]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Bacteria/Escherichia_coli_O157_H7_uid57781/NC_002127.fna -P data\n",
    "!wget ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Bacteria/Escherichia_coli_O157_H7_uid57781/NC_002127.faa -P data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <a id='2.0'>Section 2.0 - Using BLAST</a>\n",
    "\n",
    "## <a id='2.1'>Section 2.1 - Setting up the Database</a>\n",
    "Now that we have our sequences downloaded we need to setup our own version of the database using another Command Line argument called `makeblastdb`. To find out more about `makeblastdb` run the cell below...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "USAGE\n",
      "  makeblastdb [-h] [-help] [-in input_file] [-input_type type]\n",
      "    -dbtype molecule_type [-title database_title] [-parse_seqids]\n",
      "    [-hash_index] [-mask_data mask_data_files] [-mask_id mask_algo_ids]\n",
      "    [-mask_desc mask_algo_descriptions] [-gi_mask]\n",
      "    [-gi_mask_name gi_based_mask_names] [-out database_name]\n",
      "    [-max_file_sz number_of_bytes] [-logfile File_Name] [-taxid TaxID]\n",
      "    [-taxid_map TaxIDMapFile] [-version]\n",
      "\n",
      "DESCRIPTION\n",
      "   Application to create BLAST databases, version 2.6.0+\n",
      "\n",
      "Use '-help' to print detailed descriptions of command line arguments\n"
     ]
    }
   ],
   "source": [
    "!makeblastdb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running this command will slice our large sequences into a series of small **words** that can be easily sorted. This will speed up the searches that we'll be performing.\n",
    "\n",
    "Now, let's actually run `makeblastdb` on our data. Here's how it looks to run `makeblastdb` on our amino acid sequence..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 01/27/2019 00:15:37\n",
      "New DB name:   /home/ubuntu/Tutorial-Series/data/NC_002127.faa\n",
      "New DB title:  data/NC_002127.faa\n",
      "Sequence type: Protein\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 3 sequences in 0.0160251 seconds.\n"
     ]
    }
   ],
   "source": [
    "!makeblastdb -in data/NC_002127.faa -dbtype 'prot' -parse_seqids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Do the same for our nucleotide sequence below..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 01/27/2019 00:15:39\n",
      "New DB name:   /home/ubuntu/Tutorial-Series/data/NC_002127.fna\n",
      "New DB title:  data/NC_002127.fna\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 1 sequences in 0.00975394 seconds.\n"
     ]
    }
   ],
   "source": [
    "!makeblastdb -in data/NC_002127.fna -dbtype 'nucl' -parse_seqids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='2.2'>Section 2.2 - A Brief Look Into the BLAST Algorithm</a>\n",
    "\n",
    "\n",
    "Before we run our search, let's go through how BLAST works. Due to the fact that there are literally billions of sequences to search through in the NCBI database, we need a fast algorithm that could search through all of these sequences and find the ones most similar to our query. BLAST is what's called a **Heuristic** algorithm. This means that it solves the problem of sequence alignment \"well enough\".\n",
    "\n",
    "If you remember from before when we set up our databases, we sliced our original sequences into tiny **words** and then sorted them. We can easily match a word to our query sequence like in the image below...\n",
    "\n",
    "\n",
    "<img src='img/tut01/word-based1.PNG'></img>\n",
    "\n",
    "After matching a word to our query we can extend the word in both directions to narrow down which sequences are most similar.\n",
    "\n",
    "<img src='img/tut01/word-based2.PNG'></img>\n",
    "\n",
    "The word-query alignment is scored using a **BLOSUM62** matrix. From the score a series of important metrics are measured.\n",
    "\n",
    "__E-value__: An expected value or chance that this hit could have occurred randomly. It has been agreed upon that E-value < 0.05 tends to carry biological significance.\n",
    "\n",
    "__% Identity__: The percentage of matching characters between the sequences.\n",
    "\n",
    "__Query Coverage__: The percentage of how well our query sequence aligned to the database sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='2.3'>Section 2.3 - Running the BLAST Search</a>\n",
    "\n",
    "When it comes to BLAST there are few different flavours depending on what you're searching for.\n",
    "\n",
    "`blastn` ~ nulceotide query searched against nucleotide database<br>\n",
    "`blastp` ~ protein query searched against protein database<br>\n",
    "`blastx` ~ translated nulceotide query searched protein database<br>\n",
    "`tblastn` ~ protein query searched against translated nucleotide database<br>\n",
    "`tblastx` ~ translated nulceotide query searched against another translated nucleotide database\n",
    "\n",
    "to see the parameters run the cell below...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "USAGE\n",
      "  blastn [-h] [-help] [-import_search_strategy filename]\n",
      "    [-export_search_strategy filename] [-task task_name] [-db database_name]\n",
      "    [-dbsize num_letters] [-gilist filename] [-seqidlist filename]\n",
      "    [-negative_gilist filename] [-entrez_query entrez_query]\n",
      "    [-db_soft_mask filtering_algorithm] [-db_hard_mask filtering_algorithm]\n",
      "    [-subject subject_input_file] [-subject_loc range] [-query input_file]\n",
      "    [-out output_file] [-evalue evalue] [-word_size int_value]\n",
      "    [-gapopen open_penalty] [-gapextend extend_penalty]\n",
      "    [-perc_identity float_value] [-qcov_hsp_perc float_value]\n",
      "    [-max_hsps int_value] [-xdrop_ungap float_value] [-xdrop_gap float_value]\n",
      "    [-xdrop_gap_final float_value] [-searchsp int_value]\n",
      "    [-sum_stats bool_value] [-penalty penalty] [-reward reward] [-no_greedy]\n",
      "    [-min_raw_gapped_score int_value] [-template_type type]\n",
      "    [-template_length int_value] [-dust DUST_options]\n",
      "    [-filtering_db filtering_database]\n",
      "    [-window_masker_taxid window_masker_taxid]\n",
      "    [-window_masker_db window_masker_db] [-soft_masking soft_masking]\n",
      "    [-ungapped] [-culling_limit int_value] [-best_hit_overhang float_value]\n",
      "    [-best_hit_score_edge float_value] [-window_size int_value]\n",
      "    [-off_diagonal_range int_value] [-use_index boolean] [-index_name string]\n",
      "    [-lcase_masking] [-query_loc range] [-strand strand] [-parse_deflines]\n",
      "    [-outfmt format] [-show_gis] [-num_descriptions int_value]\n",
      "    [-num_alignments int_value] [-line_length line_length] [-html]\n",
      "    [-max_target_seqs num_sequences] [-num_threads int_value] [-remote]\n",
      "    [-version]\n",
      "\n",
      "DESCRIPTION\n",
      "   Nucleotide-Nucleotide BLAST 2.6.0+\n",
      "\n",
      "Use '-help' to print detailed descriptions of command line arguments\n"
     ]
    }
   ],
   "source": [
    "!blastn -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now on to the fun part! Let's search `virus_protein.fasta` against our downloaded E. coli genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TBLASTN 2.6.0+\n",
      "\n",
      "\n",
      "Reference: Stephen F. Altschul, Thomas L. Madden, Alejandro A.\n",
      "Schaffer, Jinghui Zhang, Zheng Zhang, Webb Miller, and David J.\n",
      "Lipman (1997), \"Gapped BLAST and PSI-BLAST: a new generation of\n",
      "protein database search programs\", Nucleic Acids Res. 25:3389-3402.\n",
      "\n",
      "\n",
      "\n",
      "Database: data/NC_002127.fna\n",
      "           1 sequences; 3,306 total letters\n",
      "\n",
      "\n",
      "\n",
      "Query= NP_049500.1 Shiga toxin 2 subunit A [Enterobacteria phage 933W]\n",
      "\n",
      "Length=319\n",
      "                                                                      Score     E\n",
      "Sequences producing significant alignments:                          (Bits)  Value\n",
      "\n",
      "NC_002127.1  Escherichia coli O157:H7 str. Sakai plasmid pOSAK1, ...  21.6    0.10 \n",
      "\n",
      "\n",
      ">NC_002127.1 Escherichia coli O157:H7 str. Sakai plasmid pOSAK1, complete \n",
      "sequence\n",
      "Length=3306\n",
      "\n",
      " Score = 21.6 bits (44),  Expect = 0.10, Method: Compositional matrix adjust.\n",
      " Identities = 16/47 (34%), Positives = 22/47 (47%), Gaps = 8/47 (17%)\n",
      " Frame = -1\n",
      "\n",
      "Query  86    FDHLRLIIEQNNLYVAGFVNTATNTFYRFSD------FTHISVPGVT  126\n",
      "             FD + LII++N LY          T  +FSD      F  +S  G+T\n",
      "Sbjct  2172  FDEIALIIKENCLYSD--TKNFEYTIPKFSDDDRANLFEFLSEEGIT  2038\n",
      "\n",
      "\n",
      " Score = 16.2 bits (30),  Expect = 4.8, Method: Compositional matrix adjust.\n",
      " Identities = 8/22 (36%), Positives = 9/22 (41%), Gaps = 0/22 (0%)\n",
      " Frame = +2\n",
      "\n",
      "Query  38    SLNSIRTEISTPLEHISQGTTS  59\n",
      "             SLN   T   TPL  +     S\n",
      "Sbjct  2612  SLNRCITSRQTPLNQLPMAAAS  2677\n",
      "\n",
      "\n",
      " Score = 15.0 bits (27),  Expect = 9.4, Method: Compositional matrix adjust.\n",
      " Identities = 5/10 (50%), Positives = 6/10 (60%), Gaps = 0/10 (0%)\n",
      " Frame = -1\n",
      "\n",
      "Query  147   RSGMQISRHS  156\n",
      "             RSG    RH+\n",
      "Sbjct  3252  RSGQDAERHN  3223\n",
      "\n",
      "\n",
      "\n",
      "Lambda      K        H        a         alpha\n",
      "   0.321    0.132    0.381    0.792     4.96 \n",
      "\n",
      "Gapped\n",
      "Lambda      K        H        a         alpha    sigma\n",
      "   0.267   0.0410    0.140     1.90     42.6     43.6 \n",
      "\n",
      "Effective search space used: 300330\n",
      "\n",
      "\n",
      "  Database: data/NC_002127.fna\n",
      "    Posted date:  Jan 27, 2019  12:15 AM\n",
      "  Number of letters in database: 3,306\n",
      "  Number of sequences in database:  1\n",
      "\n",
      "\n",
      "\n",
      "Matrix: BLOSUM62\n",
      "Gap Penalties: Existence: 11, Extension: 1\n",
      "Neighboring words threshold: 13\n",
      "Window for multiple hits: 40\n"
     ]
    }
   ],
   "source": [
    "!tblastn -query virus_protein.fasta -db data/NC_002127.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We should see that our viral protein actually has a hit when we search against our bacterial E. coli genome! Now since we're curious, why don't we search for similar proteins found between these two genomes...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TBLASTX 2.6.0+\n",
      "\n",
      "\n",
      "Reference: Stephen F. Altschul, Thomas L. Madden, Alejandro A.\n",
      "Schaffer, Jinghui Zhang, Zheng Zhang, Webb Miller, and David J.\n",
      "Lipman (1997), \"Gapped BLAST and PSI-BLAST: a new generation of\n",
      "protein database search programs\", Nucleic Acids Res. 25:3389-3402.\n",
      "\n",
      "\n",
      "\n",
      "Database: data/NC_002127.fna\n",
      "           1 sequences; 3,306 total letters\n",
      "\n",
      "\n",
      "\n",
      "Query= NC_000924.1 Enterobacteria phage 933W, complete genome\n",
      "\n",
      "Length=61670\n",
      "                                                                      Score     E\n",
      "Sequences producing significant alignments:                          (Bits)  Value  N\n",
      "\n",
      "NC_002127.1  Escherichia coli O157:H7 str. Sakai plasmid pOSAK1, ...  29.0    0.081  1\n",
      "\n",
      "\n",
      ">NC_002127.1 Escherichia coli O157:H7 str. Sakai plasmid pOSAK1, complete \n",
      "sequence\n",
      "Length=3306\n",
      "\n",
      " Score = 29.0 bits (57),  Expect = 0.081\n",
      " Identities = 15/46 (33%), Positives = 22/46 (48%), Gaps = 0/46 (0%)\n",
      " Frame = -2/+2\n",
      "\n",
      "Query  30472  EGNSCFISSRCSVSLLMYHCFCSSVSVMFIAFSTAEKYVIFSRDSF  30335\n",
      "              E +SCF+ S   +S+L     C+S S++    ST       S  SF\n",
      "Sbjct  1727   EHDSCFLISSWILSMLFGSLLCNSDSIVSFFMSTLNSLFFCSTISF  1864\n",
      "\n",
      "\n",
      " Score = 28.1 bits (55),  Expect = 0.15\n",
      " Identities = 11/25 (44%), Positives = 12/25 (48%), Gaps = 0/25 (0%)\n",
      " Frame = -3/+1\n",
      "\n",
      "Query  30381  PSQPQKNTSFFPVTASARFPALWRC  30307\n",
      "              P  P +N SF  V     F A WRC\n",
      "Sbjct  856    PVYPGENISFCGVCVRIAFSAFWRC  930\n",
      "\n",
      "\n",
      " Score = 26.3 bits (51),  Expect = 0.54\n",
      " Identities = 7/11 (64%), Positives = 9/11 (82%), Gaps = 0/11 (0%)\n",
      " Frame = -1/+3\n",
      "\n",
      "Query  54314  SHHWWPSAQYR  54282\n",
      "              SH WWP+ Q+R\n",
      "Sbjct  24     SHTWWPARQHR  56\n",
      "\n",
      "\n",
      " Score = 26.3 bits (51),  Expect = 0.54\n",
      " Identities = 11/25 (44%), Positives = 15/25 (60%), Gaps = 0/25 (0%)\n",
      " Frame = +1/-1\n",
      "\n",
      "Query  59839  FCAAPSSTLSFPTVSAISSFSWSRR  59913\n",
      "              FC APS   SF T S++S  ++  R\n",
      "Sbjct  627    FCNAPSRITSFSTRSSVSELTFLPR  553\n",
      "\n",
      "\n",
      " Score = 24.4 bits (47),  Expect = 1.9\n",
      " Identities = 8/22 (36%), Positives = 15/22 (68%), Gaps = 0/22 (0%)\n",
      " Frame = -3/+3\n",
      "\n",
      "Query  55824  CSAFRFITAVCTSRMKLLTRNS  55759\n",
      "              CS+F F+ +VC  R  +++R +\n",
      "Sbjct  135    CSSFCFLRSVCHLRRVIISRTT  200\n",
      "\n",
      "\n",
      " Score = 24.4 bits (47),  Expect = 1.9\n",
      " Identities = 8/10 (80%), Positives = 9/10 (90%), Gaps = 0/10 (0%)\n",
      " Frame = -2/+1\n",
      "\n",
      "Query  12793  VSYSWFRMKK  12764\n",
      "              VSYSWF MK+\n",
      "Sbjct  1438   VSYSWFHMKQ  1467\n",
      "\n",
      "\n",
      " Score = 24.0 bits (46),  Expect = 2.7\n",
      " Identities = 9/24 (38%), Positives = 13/24 (54%), Gaps = 0/24 (0%)\n",
      " Frame = +3/+2\n",
      "\n",
      "Query  61260  KKAAKLSGAETAYLFSAVSMR*LW  61331\n",
      "              ++  +L G    YL+S V  R LW\n",
      "Sbjct  2897   ERTRELPGGNAWYLYSPVGFRHLW  2968\n",
      "\n",
      "\n",
      " Score = 23.5 bits (45),  Expect = 3.6\n",
      " Identities = 8/22 (36%), Positives = 11/22 (50%), Gaps = 0/22 (0%)\n",
      " Frame = -1/-1\n",
      "\n",
      "Query  50327  HRMEPPSQHQRLHLRSLRQPRP  50262\n",
      "              HR+ PP +H +    S R   P\n",
      "Sbjct  3015   HRLRPPDEHHKNRRSSQRWRNP  2950\n",
      "\n",
      "\n",
      " Score = 23.5 bits (45),  Expect = 3.6\n",
      " Identities = 10/22 (45%), Positives = 12/22 (55%), Gaps = 0/22 (0%)\n",
      " Frame = -2/-1\n",
      "\n",
      "Query  61150  PDRQLLTGGYKLNK*GGHHDQS  61085\n",
      "              PDR  L+G Y+L      HD S\n",
      "Sbjct  2745   PDRCALSGNYRLESNPVRHDLS  2680\n",
      "\n",
      "\n",
      " Score = 23.5 bits (45),  Expect = 3.6\n",
      " Identities = 8/15 (53%), Positives = 13/15 (87%), Gaps = 0/15 (0%)\n",
      " Frame = -3/+2\n",
      "\n",
      "Query  31284  LVSFFISEPSSVWFC  31240\n",
      "              +VSFF+S  +S++FC\n",
      "Sbjct  1805   IVSFFMSTLNSLFFC  1849\n",
      "\n",
      "\n",
      " Score = 23.5 bits (45),  Expect = 3.6\n",
      " Identities = 9/18 (50%), Positives = 12/18 (67%), Gaps = 0/18 (0%)\n",
      " Frame = +3/+3\n",
      "\n",
      "Query  7467  VLGQAFFHVVCQFCLFLF  7520\n",
      "             VL ++ F  +CQ C FLF\n",
      "Sbjct  2232  VLLRSLFLYMCQICEFLF  2285\n",
      "\n",
      "\n",
      " Score = 23.5 bits (45),  Expect = 3.6\n",
      " Identities = 10/16 (63%), Positives = 10/16 (63%), Gaps = 0/16 (0%)\n",
      " Frame = +3/-3\n",
      "\n",
      "Query  32247  CFRWFLSGDTERIKPG  32294\n",
      "              CFR FLSG     KPG\n",
      "Sbjct  763    CFRQFLSGKLICHKPG  716\n",
      "\n",
      "\n",
      " Score = 23.1 bits (44),  Expect = 5.0\n",
      " Identities = 7/19 (37%), Positives = 15/19 (79%), Gaps = 0/19 (0%)\n",
      " Frame = +1/+1\n",
      "\n",
      "Query  3139  CSVVALIVINAGNVLAVVC  3195\n",
      "             CS++ L+ ++  +VL+V+C\n",
      "Sbjct  118   CSLLLLVHLSVFSVLSVIC  174\n",
      "\n",
      "\n",
      " Score = 23.1 bits (44),  Expect = 5.0\n",
      " Identities = 8/16 (50%), Positives = 14/16 (88%), Gaps = 0/16 (0%)\n",
      " Frame = -3/-1\n",
      "\n",
      "Query  49458  VTAFLPSGRVSMTSSA  49411\n",
      "              ++AFLPSG V+++ S+\n",
      "Sbjct  2865   LSAFLPSGSVALSHSS  2818\n",
      "\n",
      "\n",
      " Score = 23.1 bits (44),  Expect = 5.0\n",
      " Identities = 7/10 (70%), Positives = 7/10 (70%), Gaps = 0/10 (0%)\n",
      " Frame = +3/-2\n",
      "\n",
      "Query  3801  FAPRWAFCKN  3830\n",
      "             FAP WA C N\n",
      "Sbjct  2786  FAPSWAVCTN  2757\n",
      "\n",
      "\n",
      " Score = 23.1 bits (44),  Expect = 5.0\n",
      " Identities = 8/11 (73%), Positives = 9/11 (82%), Gaps = 0/11 (0%)\n",
      " Frame = -3/+3\n",
      "\n",
      "Query  18894  CLFIFNFLH*R  18862\n",
      "              C F+F FLH*R\n",
      "Sbjct  2271   CEFLFRFLH*R  2303\n",
      "\n",
      "\n",
      " Score = 23.1 bits (44),  Expect = 5.0\n",
      " Identities = 10/26 (38%), Positives = 13/26 (50%), Gaps = 0/26 (0%)\n",
      " Frame = +3/+2\n",
      "\n",
      "Query  5730  ISDNRKAGKYPASLCNGVDESDRAYP  5807\n",
      "             + + RKA +YP S   G  E  R  P\n",
      "Sbjct  2840  LPEGRKADRYPVSGRVGTGERTRELP  2917\n",
      "\n",
      "\n",
      " Score = 23.1 bits (44),  Expect = 5.0\n",
      " Identities = 9/20 (45%), Positives = 13/20 (65%), Gaps = 0/20 (0%)\n",
      " Frame = +1/+1\n",
      "\n",
      "Query  9508  RPTSIRQ*SRRPRRFTCDIF  9567\n",
      "             RP+   Q*+R+ +   CDIF\n",
      "Sbjct  3151  RPSVASQ*ARKRKSALCDIF  3210\n",
      "\n",
      "\n",
      " Score = 22.6 bits (43),  Expect = 6.9\n",
      " Identities = 7/8 (88%), Positives = 8/8 (100%), Gaps = 0/8 (0%)\n",
      " Frame = +2/-2\n",
      "\n",
      "Query  2351  PFSLRESW  2374\n",
      "             PFSLRE+W\n",
      "Sbjct  2858  PFSLREAW  2835\n",
      "\n",
      "\n",
      " Score = 22.6 bits (43),  Expect = 6.9\n",
      " Identities = 7/19 (37%), Positives = 13/19 (68%), Gaps = 0/19 (0%)\n",
      " Frame = -3/+2\n",
      "\n",
      "Query  21618  YVQEGSISLSVYYLTKRHN  21562\n",
      "              YV+  SIS S++++   H+\n",
      "Sbjct  1679   YVE*SSISTSIFFIISEHD  1735\n",
      "\n",
      "\n",
      " Score = 22.6 bits (43),  Expect = 6.9\n",
      " Identities = 7/13 (54%), Positives = 9/13 (69%), Gaps = 0/13 (0%)\n",
      " Frame = +1/+3\n",
      "\n",
      "Query  56095  RCMEFVP*PCRGC  56133\n",
      "              R  +F+P*PC  C\n",
      "Sbjct  1941   RACKFLP*PCHHC  1979\n",
      "\n",
      "\n",
      " Score = 22.1 bits (42),  Expect = 9.5\n",
      " Identities = 10/29 (34%), Positives = 13/29 (45%), Gaps = 0/29 (0%)\n",
      " Frame = -2/+2\n",
      "\n",
      "Query  44779  MLSFCCMCR*IYTGNPGDCTHNQCCYTLH  44693\n",
      "              MLS  C C  I+  +P   +   C Y  H\n",
      "Sbjct  110    MLSAVCCCLFIFLFSPFCLSSASCDYIAH  196\n",
      "\n",
      "\n",
      " Score = 22.1 bits (42),  Expect = 9.5\n",
      " Identities = 7/23 (30%), Positives = 13/23 (57%), Gaps = 0/23 (0%)\n",
      " Frame = +1/+3\n",
      "\n",
      "Query  5473  ILLLSCRVSLVCFCFRRHHFQLL  5541\n",
      "             +++ +CR   +C C+  H   LL\n",
      "Sbjct  285   LVIRTCRQDRMCNCYTAHTCTLL  353\n",
      "\n",
      "\n",
      " Score = 22.1 bits (42),  Expect = 9.5\n",
      " Identities = 7/28 (25%), Positives = 18/28 (64%), Gaps = 0/28 (0%)\n",
      " Frame = +2/-1\n",
      "\n",
      "Query  58040  SGLYQFHPQFLLCLSEKHSDRPICCHRY  58123\n",
      "              + L++F  +  + ++E +++ P C H+Y\n",
      "Sbjct  2076   ANLFEFLSEEGITITEDNNNDPNCKHQY  1993\n",
      "\n",
      "\n",
      " Score = 22.1 bits (42),  Expect = 9.5\n",
      " Identities = 7/16 (44%), Positives = 12/16 (75%), Gaps = 0/16 (0%)\n",
      " Frame = -1/+3\n",
      "\n",
      "Query  28025  LIYFWVPSSCFCIVIS  27978\n",
      "              ++YF+V  S  C+V+S\n",
      "Sbjct  1833   ILYFFVRQSLSCLVLS  1880\n",
      "\n",
      "\n",
      " Score = 22.1 bits (42),  Expect = 9.5\n",
      " Identities = 5/6 (83%), Positives = 5/6 (83%), Gaps = 0/6 (0%)\n",
      " Frame = -2/+3\n",
      "\n",
      "Query  59962  PCWHCW  59945\n",
      "              PC HCW\n",
      "Sbjct  1965   PCHHCW  1982\n",
      "\n",
      "\n",
      " Score = 22.1 bits (42),  Expect = 9.5\n",
      " Identities = 5/6 (83%), Positives = 5/6 (83%), Gaps = 0/6 (0%)\n",
      " Frame = -3/+3\n",
      "\n",
      "Query  4638  PCWHCW  4621\n",
      "             PC HCW\n",
      "Sbjct  1965  PCHHCW  1982\n",
      "\n",
      "\n",
      " Score = 22.1 bits (42),  Expect = 9.5\n",
      " Identities = 12/34 (35%), Positives = 19/34 (56%), Gaps = 0/34 (0%)\n",
      " Frame = +1/-1\n",
      "\n",
      "Query  12787  TILKHLQKLSVQP*YVLKVRKRIFLCKILMSRLK  12888\n",
      "              T ++HL   S Q  +  K+RK + +   L SRL+\n",
      "Sbjct  1158   TSIRHLMARSFQVTH*TKLRKLMNIMIYLSSRLR  1057\n",
      "\n",
      "\n",
      " Score = 22.1 bits (42),  Expect = 9.5\n",
      " Identities = 8/20 (40%), Positives = 14/20 (70%), Gaps = 0/20 (0%)\n",
      " Frame = +1/-1\n",
      "\n",
      "Query  34957  ESAAHEV*PGILSQNASGDV  35016\n",
      "              ++A  +V PG+  +NASG +\n",
      "Sbjct  894    DTAKRDVFPGVNRKNASGSL  835\n",
      "\n",
      "\n",
      " Score = 22.1 bits (42),  Expect = 9.5\n",
      " Identities = 6/8 (75%), Positives = 7/8 (88%), Gaps = 0/8 (0%)\n",
      " Frame = -3/-1\n",
      "\n",
      "Query  28080  HNNCMSCP  28057\n",
      "              +NNC SCP\n",
      "Sbjct  330    YNNCTSCP  307\n",
      "\n",
      "\n",
      " Score = 22.1 bits (42),  Expect = 9.5\n",
      " Identities = 11/35 (31%), Positives = 14/35 (40%), Gaps = 0/35 (0%)\n",
      " Frame = +2/+2\n",
      "\n",
      "Query  59561  A*RIPEPCFRRTPEAIAQPFAHQCERHKCDRHQIA  59665\n",
      "              A*R+ E    R P     P+A  C    C    I+\n",
      "Sbjct  3158   A*RVSERGSGRAPYVTFSPYALLCRSASCPERYIS  3262\n",
      "\n",
      "\n",
      "\n",
      "Lambda      K        H\n",
      "   0.318    0.134    0.401 \n",
      "\n",
      "\n",
      "Effective search space used: 21939087\n",
      "\n",
      "\n",
      "  Database: data/NC_002127.fna\n",
      "    Posted date:  Jan 27, 2019  12:15 AM\n",
      "  Number of letters in database: 3,306\n",
      "  Number of sequences in database:  1\n",
      "\n",
      "\n",
      "\n",
      "Matrix: BLOSUM62\n",
      "Neighboring words threshold: 13\n",
      "Window for multiple hits: 40\n"
     ]
    }
   ],
   "source": [
    "!tblastx -query virus_genome.fasta -db data/NC_002127.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From what we're seeing, there appears to be a ton of matches for these viral proteins inside of our bacterial genome! How could this have happened?\n",
    "\n",
    "Here's a link to our slides that describe what a bacteriophage is and how bacteria can undergo lysogenic conversion.\n",
    "\n",
    "https://docs.google.com/presentation/d/1g2OVAMlUyMv_oTlyTEUz5ZMwJ1QcHiuQUI2GsORd1iw/edit?usp=sharing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
