{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Before you begin:\n",
    "Run the following line and ensure that it prints out something like\n",
    "```\n",
    "/home/ubuntu/<Your quest ID>/Tutorial-Series\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/Tutorial-Series\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of Contents\n",
    "- <a href='#1.0'>Section 1 - Introduction and Recap</a>\n",
    "- <a href='#2.0'>Section 2 - Intro to Annotation</a>\n",
    "    - <a href='#2.1'>Section 2.1 - What is Annotation?</a>\n",
    "- <a href='#3.0'>Section 3 - Types of Annotation</a>\n",
    "    - <a href='#3.1'>Section 3.1 - Structural Annotation</a>\n",
    "    - <a href='#3.2'>Section 3.2 - Functional Annotation</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial 3 - Genome Annotation\n",
    "\n",
    "## <a id='1.0'>Section 1 - Introduction and Recap</a>\n",
    "\n",
    "In the last tutorial we touched on sequencing, and how to perform genome assemblies. We also learned about the computational aspects of genome assemblies. Afterwards, we ran a BLAST search on our assembled contigs to identify which strain of E. coli they belong to. We eventually found out that the contigs belonged to a strain called K12.\n",
    "\n",
    "In this tutorial we're going to try and describe the functions in the K12 genome through annotation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='2.0'>Section 2 - Intro to Annotation</a>\n",
    "\n",
    "### <a id='2.0'>Section 2.1 - What is Annotation?</a>\n",
    "\n",
    "Annotation refers to the process of adding supplemental information onto a sequence (nucleotide or protein) so that we can better understand the underlying function/purpose of said sequence. This isn't an easy task due to the tremendous amount of processes needed to keep an organism alive. There's also the problem of mutations, a sequence may or may not change is function due to mutations, so we need to make sure the algorithms we use are robust enough to take mutations into consideration.\n",
    "\n",
    "To ensure that the programs we use for annotation can account for these mutations, databases for protein families such as <a href='https://pfam.xfam.org/'>Pfam</a> are maintained. Many of the annotation software in use, search our sequences against some advanced statistical models called HMMs, which are also maintained on <a href='https://pfam.xfam.org/'>Pfam</a>. We won't go over HMMs in this tutorial, but you should know that they're an important method for annotation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='3.0'>Section 3 - Types of Annotation</a>\n",
    "\n",
    "### <a id='3.1'> Section 3.1 - Structural Annotation</a>\n",
    "\n",
    "Structural annotation refers to locating all of the coding regions in your sequence and trying to identify the genes in those coding regions. In this tutorial, we'll be using one of the most popular programs for prokaryotic genome annotation called PROKKA.\n",
    "\n",
    "PROKKA is a \"Pipeline\" or just a series of other programs that are strung together to process our data.\n",
    "\n",
    "<img src='img/tut03/prokka.jpg'></img>\n",
    "\n",
    "We're going to use PROKKA to annotate our `K12.fasta` file and we're going to save it in our `data/tut03/K12_annot` folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[23:06:08] Appending to PATH: /home/ubuntu/prokka/bin/../binaries/linux\n",
      "[23:06:08] Appending to PATH: /home/ubuntu/prokka/bin/../binaries/linux/../common\n",
      "[23:06:08] Appending to PATH: /home/ubuntu/prokka/bin\n",
      "[23:06:08] Cleaning databases in /home/ubuntu/prokka/db\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Archaea/sprot.phd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Archaea/sprot.phi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Archaea/sprot.phr\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Archaea/sprot.pin\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Archaea/sprot.pog\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Archaea/sprot.psd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Archaea/sprot.psi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Archaea/sprot.psq\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/AMR.phd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/AMR.phi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/AMR.phr\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/AMR.pin\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/AMR.pog\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/AMR.psd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/AMR.psi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/AMR.psq\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/IS.phd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/IS.phi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/IS.phr\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/IS.pin\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/IS.pog\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/IS.psd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/IS.psi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/IS.psq\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/sprot.phd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/sprot.phi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/sprot.phr\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/sprot.pin\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/sprot.pog\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/sprot.psd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/sprot.psi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Bacteria/sprot.psq\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Mitochondria/sprot.phd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Mitochondria/sprot.phi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Mitochondria/sprot.phr\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Mitochondria/sprot.pin\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Mitochondria/sprot.pog\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Mitochondria/sprot.psd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Mitochondria/sprot.psi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Mitochondria/sprot.psq\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Viruses/sprot.phd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Viruses/sprot.phi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Viruses/sprot.phr\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Viruses/sprot.pin\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Viruses/sprot.pog\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Viruses/sprot.psd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Viruses/sprot.psi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/kingdom/Viruses/sprot.psq\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Enterococcus.phd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Enterococcus.phi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Enterococcus.phr\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Enterococcus.pin\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Enterococcus.pog\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Enterococcus.psd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Enterococcus.psi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Enterococcus.psq\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Escherichia.phd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Escherichia.phi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Escherichia.phr\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Escherichia.pin\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Escherichia.pog\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Escherichia.psd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Escherichia.psi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Escherichia.psq\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Staphylococcus.phd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Staphylococcus.phi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Staphylococcus.phr\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Staphylococcus.pin\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Staphylococcus.pog\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Staphylococcus.psd\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Staphylococcus.psi\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/genus/Staphylococcus.psq\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/hmm/HAMAP.hmm.h3f\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/hmm/HAMAP.hmm.h3i\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/hmm/HAMAP.hmm.h3m\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/hmm/HAMAP.hmm.h3p\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/cm/Bacteria.i1f\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/cm/Bacteria.i1i\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/cm/Bacteria.i1m\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/cm/Bacteria.i1p\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/cm/Viruses.i1f\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/cm/Viruses.i1i\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/cm/Viruses.i1m\n",
      "[23:06:08] Deleting unwanted file: /home/ubuntu/prokka/db/cm/Viruses.i1p\n",
      "[23:06:08] Cleaning complete.\n",
      "[23:06:08] Looking for 'makeblastdb' - found /usr/bin/makeblastdb\n",
      "[23:06:08] Determined makeblastdb version is 2.6\n",
      "[23:06:08] Making kingdom BLASTP database: /home/ubuntu/prokka/db/kingdom/Archaea/sprot\n",
      "[23:06:08] Running: makeblastdb -hash_index -dbtype prot -in \\/home\\/ubuntu\\/prokka\\/db\\/kingdom\\/Archaea\\/sprot -logfile /dev/null\n",
      "[23:06:08] Making kingdom BLASTP database: /home/ubuntu/prokka/db/kingdom/Bacteria/sprot\n",
      "[23:06:08] Running: makeblastdb -hash_index -dbtype prot -in \\/home\\/ubuntu\\/prokka\\/db\\/kingdom\\/Bacteria\\/sprot -logfile /dev/null\n",
      "[23:06:09] Making kingdom BLASTP database: /home/ubuntu/prokka/db/kingdom/Mitochondria/sprot\n",
      "[23:06:09] Running: makeblastdb -hash_index -dbtype prot -in \\/home\\/ubuntu\\/prokka\\/db\\/kingdom\\/Mitochondria\\/sprot -logfile /dev/null\n",
      "[23:06:09] Making kingdom BLASTP database: /home/ubuntu/prokka/db/kingdom/Viruses/sprot\n",
      "[23:06:09] Running: makeblastdb -hash_index -dbtype prot -in \\/home\\/ubuntu\\/prokka\\/db\\/kingdom\\/Viruses\\/sprot -logfile /dev/null\n",
      "[23:06:09] Making kingdom BLASTP database: /home/ubuntu/prokka/db/kingdom/Bacteria/IS\n",
      "[23:06:09] Running: makeblastdb -hash_index -dbtype prot -in \\/home\\/ubuntu\\/prokka\\/db\\/kingdom\\/Bacteria\\/IS -logfile /dev/null\n",
      "[23:06:09] Making kingdom BLASTP database: /home/ubuntu/prokka/db/kingdom/Bacteria/AMR\n",
      "[23:06:09] Running: makeblastdb -hash_index -dbtype prot -in \\/home\\/ubuntu\\/prokka\\/db\\/kingdom\\/Bacteria\\/AMR -logfile /dev/null\n",
      "[23:06:10] Making genus BLASTP database: /home/ubuntu/prokka/db/genus/Enterococcus\n",
      "[23:06:10] Running: makeblastdb -hash_index -dbtype prot -in \\/home\\/ubuntu\\/prokka\\/db\\/genus\\/Enterococcus -logfile /dev/null\n",
      "[23:06:10] Making genus BLASTP database: /home/ubuntu/prokka/db/genus/Escherichia\n",
      "[23:06:10] Running: makeblastdb -hash_index -dbtype prot -in \\/home\\/ubuntu\\/prokka\\/db\\/genus\\/Escherichia -logfile /dev/null\n",
      "[23:06:10] Making genus BLASTP database: /home/ubuntu/prokka/db/genus/Staphylococcus\n",
      "[23:06:10] Running: makeblastdb -hash_index -dbtype prot -in \\/home\\/ubuntu\\/prokka\\/db\\/genus\\/Staphylococcus -logfile /dev/null\n",
      "[23:06:10] Looking for 'hmmpress' - found /usr/bin/hmmpress\n",
      "[23:06:10] Determined hmmpress version is 3.1\n",
      "[23:06:10] Pressing HMM database: /home/ubuntu/prokka/db/hmm/HAMAP.hmm\n",
      "[23:06:10] Running: hmmpress \\/home\\/ubuntu\\/prokka\\/db\\/hmm\\/HAMAP\\.hmm\n",
      "Working...    done.\n",
      "Pressed and indexed 1463 HMMs (1463 names).\n",
      "Models pressed into binary file:   /home/ubuntu/prokka/db/hmm/HAMAP.hmm.h3m\n",
      "SSI index for binary model file:   /home/ubuntu/prokka/db/hmm/HAMAP.hmm.h3i\n",
      "Profiles (MSV part) pressed into:  /home/ubuntu/prokka/db/hmm/HAMAP.hmm.h3f\n",
      "Profiles (remainder) pressed into: /home/ubuntu/prokka/db/hmm/HAMAP.hmm.h3p\n",
      "[23:06:11] Looking for 'cmpress' - found /usr/bin/cmpress\n",
      "[23:06:11] Determined cmpress version is 1.1\n",
      "[23:06:11] Pressing CM database: /home/ubuntu/prokka/db/cm/Viruses\n",
      "[23:06:11] Running: cmpress \\/home\\/ubuntu\\/prokka\\/db\\/cm\\/Viruses\n",
      "Working...    done.\n",
      "Pressed and indexed 142 CMs and p7 HMM filters (142 names and 142 accessions).\n",
      "Covariance models and p7 filters pressed into binary file:  /home/ubuntu/prokka/db/cm/Viruses.i1m\n",
      "SSI index for binary covariance model file:                 /home/ubuntu/prokka/db/cm/Viruses.i1i\n",
      "Optimized p7 filter profiles (MSV part)  pressed into:      /home/ubuntu/prokka/db/cm/Viruses.i1f\n",
      "Optimized p7 filter profiles (remainder) pressed into:      /home/ubuntu/prokka/db/cm/Viruses.i1p\n",
      "[23:06:12] Pressing CM database: /home/ubuntu/prokka/db/cm/Bacteria\n",
      "[23:06:12] Running: cmpress \\/home\\/ubuntu\\/prokka\\/db\\/cm\\/Bacteria\n",
      "Working...    done.\n",
      "Pressed and indexed 1027 CMs and p7 HMM filters (1027 names and 1027 accessions).\n",
      "Covariance models and p7 filters pressed into binary file:  /home/ubuntu/prokka/db/cm/Bacteria.i1m\n",
      "SSI index for binary covariance model file:                 /home/ubuntu/prokka/db/cm/Bacteria.i1i\n",
      "Optimized p7 filter profiles (MSV part)  pressed into:      /home/ubuntu/prokka/db/cm/Bacteria.i1f\n",
      "Optimized p7 filter profiles (remainder) pressed into:      /home/ubuntu/prokka/db/cm/Bacteria.i1p\n",
      "[23:06:12] Looking for databases in: /home/ubuntu/prokka/db\n",
      "[23:06:12] * Kingdoms: Archaea Bacteria Bacteria Bacteria Mitochondria Viruses\n",
      "[23:06:12] * Genera: Enterococcus Escherichia Staphylococcus\n",
      "[23:06:12] * HMMs: HAMAP\n",
      "[23:06:12] * CMs: Bacteria Viruses\n"
     ]
    }
   ],
   "source": [
    "!prokka K12.fasta --setupdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[23:06:15] This is prokka 1.13.3\n",
      "[23:06:15] Written by Torsten Seemann <torsten.seemann@gmail.com>\n",
      "[23:06:15] Homepage is https://github.com/tseemann/prokka\n",
      "[23:06:15] Local time is Thu Feb  7 23:06:15 2019\n",
      "[23:06:15] You are ubuntu\n",
      "[23:06:15] Operating system is linux\n",
      "[23:06:15] You have BioPerl 1.007002\n",
      "[23:06:15] System has 64 cores.\n",
      "[23:06:15] Will use maximum of 8 cores.\n",
      "[23:06:15] Annotating as >>> Bacteria <<<\n",
      "[23:06:15] Generating locus_tag from 'K12.fasta' contents.\n",
      "[23:06:15] Setting --locustag HOICKCKO from MD5 182c4c48740aa62f1b6ca3c03afd663a\n",
      "[23:06:15] Re-using existing --outdir data/tut03/K12_annot\n",
      "[23:06:15] Using filename prefix: PROKKA_02072019.XXX\n",
      "[23:06:15] Setting HMMER_NCPU=1\n",
      "[23:06:15] Writing log to: data/tut03/K12_annot/PROKKA_02072019.log\n",
      "[23:06:15] Command: /home/ubuntu/prokka/bin/prokka K12.fasta --outdir data/tut03/K12_annot --force\n",
      "[23:06:15] Appending to PATH: /home/ubuntu/prokka/bin/../binaries/linux\n",
      "[23:06:15] Appending to PATH: /home/ubuntu/prokka/bin/../binaries/linux/../common\n",
      "[23:06:15] Appending to PATH: /home/ubuntu/prokka/bin\n",
      "[23:06:15] Looking for 'aragorn' - found /home/ubuntu/prokka/bin/../binaries/linux/aragorn\n",
      "[23:06:15] Determined aragorn version is 1.2\n",
      "[23:06:15] Looking for 'blastp' - found /usr/bin/blastp\n",
      "[23:06:15] Determined blastp version is 2.6\n",
      "[23:06:15] Looking for 'cmpress' - found /usr/bin/cmpress\n",
      "[23:06:15] Determined cmpress version is 1.1\n",
      "[23:06:15] Looking for 'cmscan' - found /usr/bin/cmscan\n",
      "[23:06:15] Determined cmscan version is 1.1\n",
      "[23:06:15] Looking for 'egrep' - found /bin/egrep\n",
      "[23:06:15] Looking for 'find' - found /usr/bin/find\n",
      "[23:06:15] Looking for 'grep' - found /bin/grep\n",
      "[23:06:15] Looking for 'hmmpress' - found /usr/bin/hmmpress\n",
      "[23:06:15] Determined hmmpress version is 3.1\n",
      "[23:06:15] Looking for 'hmmscan' - found /usr/bin/hmmscan\n",
      "[23:06:15] Determined hmmscan version is 3.1\n",
      "[23:06:15] Looking for 'java' - found /usr/bin/java\n",
      "[23:06:15] Looking for 'less' - found /usr/bin/less\n",
      "[23:06:15] Looking for 'makeblastdb' - found /usr/bin/makeblastdb\n",
      "[23:06:15] Determined makeblastdb version is 2.6\n",
      "[23:06:15] Looking for 'minced' - found /home/ubuntu/prokka/bin/../binaries/linux/../common/minced\n",
      "[23:06:16] Determined minced version is 2.0\n",
      "[23:06:16] Looking for 'parallel' - found /home/ubuntu/prokka/bin/../binaries/linux/../common/parallel\n",
      "[23:06:16] Determined parallel version is 20150122\n",
      "[23:06:16] Looking for 'prodigal' - found /home/ubuntu/prokka/bin/../binaries/linux/prodigal\n",
      "[23:06:16] Determined prodigal version is 2.6\n",
      "[23:06:16] Looking for 'prokka-genbank_to_fasta_db' - found /home/ubuntu/prokka/bin/prokka-genbank_to_fasta_db\n",
      "[23:06:16] Looking for 'sed' - found /bin/sed\n",
      "[23:06:16] Looking for 'tbl2asn' - found /home/ubuntu/prokka/bin/../binaries/linux/tbl2asn\n",
      "[23:06:16] Determined tbl2asn version is 25.6\n",
      "[23:06:16] Using genetic code table 11.\n",
      "[23:06:16] Loading and checking input file: K12.fasta\n",
      "[23:06:16] Wrote 1 contigs totalling 4641652 bp.\n",
      "[23:06:16] Predicting tRNAs and tmRNAs\n",
      "[23:06:16] Running: aragorn -l -gc11  -w data\\/tut03\\/K12_annot\\/PROKKA_02072019\\.fna\n",
      "[23:06:18] 1 tRNA-Ile [225381,225457] 35 (gat)\n",
      "[23:06:18] 2 tRNA-Ala [225500,225575] 34 (tgc)\n",
      "[23:06:18] 3 tRNA-Asp [228928,229004] 35 (gtc)\n",
      "[23:06:18] 4 tRNA-Asp [236931,237007] 35 (gtc)\n",
      "[23:06:18] 5 tRNA-Thr [262871,262946] 34 (cgt)\n",
      "[23:06:18] 6 tRNA-Ser [345334,345414] 38 (gga)\n",
      "[23:06:18] 7 tRNA-Arg [564723,564799] 35 (tct)\n",
      "[23:06:18] 8 tRNA-Gln c[696430,696504] 33 (ctg)\n",
      "[23:06:18] 9 tRNA-Gln c[696542,696616] 33 (ctg)\n",
      "[23:06:18] 10 tRNA-Met c[696664,696740] 35 (cat)\n",
      "[23:06:18] 11 tRNA-Gln c[696756,696830] 33 (ttg)\n",
      "[23:06:18] 12 tRNA-Gln c[696865,696939] 33 (ttg)\n",
      "[23:06:18] 13 tRNA-Leu c[696963,697047] 35 (tag)\n",
      "[23:06:18] 14 tRNA-Met c[697057,697133] 35 (cat)\n",
      "[23:06:18] 15 tRNA-Lys [780554,780629] 34 (ttt)\n",
      "[23:06:18] 16 tRNA-Val [780765,780840] 34 (tac)\n",
      "[23:06:18] 17 tRNA-Lys [780843,780918] 34 (ttt)\n",
      "[23:06:18] 18 tRNA-Val [781068,781143] 34 (tac)\n",
      "[23:06:18] 19 tRNA-Lys [781147,781222] 34 (ttt)\n",
      "[23:06:18] 20 tRNA-Lys [781369,781444] 34 (ttt)\n",
      "[23:06:18] 21 tRNA-Lys [781577,781652] 34 (ttt)\n",
      "[23:06:18] 22 tRNA-Ser c[925884,925971] 35 (gga)\n",
      "[23:06:18] 23 tRNA-Ser c[1031625,1031712] 35 (tga)\n",
      "[23:06:18] 24 tRNA-Ser c[1097565,1097652] 35 (gga)\n",
      "[23:06:18] 25 tRNA-Tyr c[1287244,1287328] 35 (gta)\n",
      "[23:06:18] 26 tRNA-Tyr c[1287538,1287622] 35 (gta)\n",
      "[23:06:18] 27 tRNA-Val [1746435,1746511] 35 (gac)\n",
      "[23:06:18] 28 tRNA-Val [1746516,1746592] 35 (gac)\n",
      "[23:06:18] 29 tRNA-Leu c[1991815,1991901] 35 (taa)\n",
      "[23:06:18] 30 tRNA-Cys c[1991914,1991987] 33 (gca)\n",
      "[23:06:18] 31 tRNA-Gly c[1992042,1992117] 34 (gcc)\n",
      "[23:06:18] 32 tRNA-Ser c[2043468,2043557] 35 (cga)\n",
      "[23:06:18] 33 tRNA-Asn [2044549,2044624] 34 (gtt)\n",
      "[23:06:18] 34 tRNA-Asn c[2058027,2058102] 34 (gtt)\n",
      "[23:06:18] 35 tRNA-Asn [2059851,2059926] 34 (gtt)\n",
      "[23:06:18] 36 tRNA-Asn [2062260,2062335] 34 (gtt)\n",
      "[23:06:18] 37 tRNA-Pro [2286211,2286287] 35 (ggg)\n",
      "[23:06:18] 38 tRNA-Arg [2466309,2466383] 34 (cct)\n",
      "[23:06:18] 39 tRNA-Ala c[2518041,2518116] 34 (ggc)\n",
      "[23:06:18] 40 tRNA-Ala c[2518156,2518231] 34 (ggc)\n",
      "[23:06:18] 41 tRNA-Val [2520931,2521006] 34 (tac)\n",
      "[23:06:18] 42 tRNA-Val [2521051,2521126] 34 (tac)\n",
      "[23:06:18] 43 tRNA-Val [2521173,2521248] 34 (tac)\n",
      "[23:06:18] 44 tRNA-Lys [2521253,2521328] 34 (ttt)\n",
      "[23:06:18] 45 tRNA-Glu c[2729369,2729444] 35 (ttc)\n",
      "[23:06:18] 46 tmRNA [2755593,2755955] 90,125 ANDENYALAA**\n",
      "[23:06:18] 47 tRNA-Met c[2785762,2785837] 34 (cat)\n",
      "[23:06:18] 48 tRNA-Arg c[2817784,2817860] 35 (acg)\n",
      "[23:06:18] 49 tRNA-Arg c[2818059,2818135] 35 (acg)\n",
      "[23:06:18] 50 tRNA-Arg c[2818198,2818274] 35 (acg)\n",
      "[23:06:18] 51 tRNA-Arg [2818458,2818529] 34 (tcg)\n",
      "[23:06:18] 52 tRNA-Arg c[2818473,2818549] 35 (acg)\n",
      "[23:06:18] 53 tRNA-Ser c[2818553,2818645] 35 (gct)\n",
      "[23:06:18] 54 tRNA-Met [2947387,2947463] 35 (cat)\n",
      "[23:06:18] 55 tRNA-Met [2947497,2947573] 35 (cat)\n",
      "[23:06:18] 56 tRNA-Met [2947607,2947683] 35 (cat)\n",
      "[23:06:18] 57 tRNA-Gly c[2998984,2999057] 33 (ccc)\n",
      "[23:06:18] 58 tRNA-Phe [3110366,3110441] 34 (gaa)\n",
      "[23:06:18] 59 tRNA-Met [3215598,3215673] 34 (cat)\n",
      "[23:06:18] 60 tRNA-Met c[3318213,3318289] 35 (cat)\n",
      "[23:06:18] 61 tRNA-Leu c[3322072,3322158] 35 (gag)\n",
      "[23:06:18] 62 tRNA-Thr c[3423580,3423655] 34 (ggt)\n",
      "[23:06:18] 63 tRNA-Ala c[3426958,3427033] 34 (tgc)\n",
      "[23:06:18] 64 tRNA-Ile c[3427076,3427152] 35 (gat)\n",
      "[23:06:18] 65 tRNA-Pro c[3708616,3708692] 35 (cgg)\n",
      "[23:06:18] 66 tRNA-SeC [3836222,3836316] 35 (tca)\n",
      "[23:06:18] 67 tRNA-Glu [3943435,3943510] 35 (ttc)\n",
      "[23:06:18] 68 tRNA-Asp [3946872,3946948] 35 (gtc)\n",
      "[23:06:18] 69 tRNA-Trp [3946957,3947032] 34 (cca)\n",
      "[23:06:18] 70 tRNA-Arg [3982375,3982451] 35 (ccg)\n",
      "[23:06:18] 71 tRNA-His [3982510,3982585] 34 (gtg)\n",
      "[23:06:18] 72 tRNA-Leu [3982606,3982692] 35 (cag)\n",
      "[23:06:18] 73 tRNA-Pro [3982735,3982811] 35 (tgg)\n",
      "[23:06:18] 74 tRNA-Ile [4037141,4037217] 35 (gat)\n",
      "[23:06:18] 75 tRNA-Ala [4037260,4037335] 34 (tgc)\n",
      "[23:06:18] 76 tRNA-Glu [4168372,4168447] 35 (ttc)\n",
      "[23:06:18] 77 tRNA-Thr [4175388,4175463] 34 (tgt)\n",
      "[23:06:18] 78 tRNA-Tyr [4175472,4175556] 35 (gta)\n",
      "[23:06:18] 79 tRNA-Gly [4175673,4175747] 34 (tcc)\n",
      "[23:06:18] 80 tRNA-Thr [4175754,4175829] 34 (ggt)\n",
      "[23:06:18] 81 tRNA-Glu [4209774,4209849] 35 (ttc)\n",
      "[23:06:18] 82 tRNA-Phe c[4362551,4362626] 34 (gaa)\n",
      "[23:06:18] 83 tRNA-Gly [4392360,4392435] 34 (gcc)\n",
      "[23:06:18] 84 tRNA-Gly [4392472,4392547] 34 (gcc)\n",
      "[23:06:18] 85 tRNA-Gly [4392583,4392658] 34 (gcc)\n",
      "[23:06:18] 86 tRNA-Leu [4496405,4496489] 35 (caa)\n",
      "[23:06:18] 87 tRNA-Leu c[4606079,4606165] 35 (cag)\n",
      "[23:06:18] 88 tRNA-Leu c[4606200,4606286] 35 (cag)\n",
      "[23:06:18] 89 tRNA-Leu c[4606315,4606401] 35 (cag)\n",
      "[23:06:18] Found 89 tRNAs\n",
      "[23:06:18] Predicting Ribosomal RNAs\n",
      "[23:06:18] You need either Barrnap or RNAmmer installed to predict rRNAs!\n",
      "[23:06:18] Skipping ncRNA search, enable with --rfam if desired.\n",
      "[23:06:18] Total of 88 tRNA + rRNA features\n",
      "[23:06:18] Searching for CRISPR repeats\n",
      "[23:06:19] CRISPR1 NC_000913.3 2877701 with 13 spacers\n",
      "[23:06:19] CRISPR2 NC_000913.3 2904014 with 7 spacers\n",
      "[23:06:19] Found 2 CRISPRs\n",
      "[23:06:19] Predicting coding sequences\n",
      "[23:06:19] Contigs total 4641652 bp, so using single mode\n",
      "[23:06:19] Running: prodigal -i data\\/tut03\\/K12_annot\\/PROKKA_02072019\\.fna -c -m -g 11 -p single -f sco -q\n",
      "[23:06:29] Excluding CDS which overlaps existing RNA (tRNA) at NC_000913.3:1287065..1287631 on + strand\n",
      "[23:06:31] Excluding CDS which overlaps existing RNA (tRNA) at NC_000913.3:2817923..2818393 on + strand\n",
      "[23:06:31] Excluding CDS which overlaps existing RNA (tRNA) at NC_000913.3:3423503..3423826 on + strand\n",
      "[23:06:33] Found 4315 CDS\n",
      "[23:06:33] Connecting features back to sequences\n",
      "[23:06:33] Not using genus-specific database. Try --usegenus to enable it.\n",
      "[23:06:33] Annotating CDS, please be patient.\n",
      "[23:06:33] Will use 8 CPUs for similarity searching.\n",
      "[23:06:36] There are still 4315 unannotated CDS left (started with 4315)\n",
      "[23:06:36] Will use blast to search against /home/ubuntu/prokka/db/kingdom/Bacteria/IS with 8 CPUs\n",
      "[23:06:36] Running: cat data\\/tut03\\/K12_annot\\/IS\\.faa | parallel --gnu --plain -j 8 --block 86334 --recstart '>' --pipe blastp -query - -db /home/ubuntu/prokka/db/kingdom/Bacteria/IS -evalue 1e-30 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > data\\/tut03\\/K12_annot\\/IS\\.blast 2> /dev/null\n",
      "[23:06:50] Deleting unwanted file: data/tut03/K12_annot/IS.faa\n",
      "[23:06:50] Deleting unwanted file: data/tut03/K12_annot/IS.blast\n",
      "[23:06:53] There are still 4239 unannotated CDS left (started with 4315)\n",
      "[23:06:53] Will use blast to search against /home/ubuntu/prokka/db/kingdom/Bacteria/AMR with 8 CPUs\n",
      "[23:06:53] Running: cat data\\/tut03\\/K12_annot\\/AMR\\.faa | parallel --gnu --plain -j 8 --block 85202 --recstart '>' --pipe blastp -query - -db /home/ubuntu/prokka/db/kingdom/Bacteria/AMR -evalue 9.99999999999999e-301 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > data\\/tut03\\/K12_annot\\/AMR\\.blast 2> /dev/null\n",
      "[23:07:01] Could not run command: cat data\\/tut03\\/K12_annot\\/AMR\\.faa | parallel --gnu --plain -j 8 --block 85202 --recstart '>' --pipe blastp -query - -db /home/ubuntu/prokka/db/kingdom/Bacteria/AMR -evalue 9.99999999999999e-301 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > data\\/tut03\\/K12_annot\\/AMR\\.blast 2> /dev/null\n"
     ]
    }
   ],
   "source": [
    "!prokka K12.fasta --outdir data/tut03/K12_annot --force"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_002695.2 Escherichia coli O157:H7 str. Sakai DNA, complete genome\n",
      "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTCTCTGACAGCAGC\n",
      "TTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAA\n",
      "TATAGGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATCCATGAAACGCATTAGCACCACC\n",
      "ATTACCACCACCATCACCACCACCATCACCATTACCATTACCACAGGTAACGGTGCGGGCTGACGCGTAC\n"
     ]
    }
   ],
   "source": [
    "!head -n 5 O157H7.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have annotated the genes in our `K12.fasta` we can extract the genes that were found using this command below and save it as `genelist.txt`...\n",
    "\n",
    "(FYI to run this command in Jupyter you need to open up a terminal window)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cut: the delimiter must be a single character\n",
      "Try 'cut --help' for more information.\n"
     ]
    }
   ],
   "source": [
    "!cut -d$'\\t' -f4 data/tut03/K12_annot/*.tsv | uniq >genelist.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='3.2'> Section 3.2 - Functional Annotation</a>\n",
    "\n",
    "Functional Annotation allows us to gauge the function of the proteins in our sequence. This is a very useful if you want to identify any unique behaviours in your sample. An example of this could be if you want to identify whether your E. coli sample is pathogenic. It could also provide you with a summary of how your bacteria behaves in a sample. \n",
    "\n",
    "To do this we're going to use the ever popular <a href='http://www.geneontology.org/'>Gene Ontology Consortium</a>.\n",
    "\n",
    "We need to download our `genelist.txt` file and copy the genes into the website."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the next tutorial, we'll take all of our annotated files and perform some comparative genomics!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
